client/allocrunner/taskrunner/template/template.go

```diff
@@ -881,30 +885,32 @@ func newRunnerConfig(config *TaskTemplateManagerConfig,
 	emptyStr := ""
 	conf.Vault.RenewToken = pointer.Of(false)
 	conf.Vault.Token = &emptyStr
-	if cc.VaultConfig != nil && cc.VaultConfig.IsEnabled() {
-		conf.Vault.Address = &cc.VaultConfig.Addr
+
+	vaultConfig := cc.GetVaultConfigs(nil)[config.VaultCluster]
+	if vaultConfig != nil && vaultConfig.IsEnabled() {
```

Comment:
`GetVaultConfigs()` in ENT does not return disabled configs, so it's not possible to differentiate between a cluster that is not configured (and that should maybe result in an error) from one that is disabled (expected, no error).

But I'm not sure if this is relevant. Tasks should only run in clients that do have the given config because of the constraint. 

In case the config does get changed and the client restarts, I think it's expected the task to just keep running?

Reply:
The task _should_ keep running but the template runner isn't forked from the Nomad agent. So if you break the Vault config while the client is down, the task is going to fail to restore. I think that's expected behavior though.

Reply:
> The task should keep running but the template runner isn't forked from the Nomad agent. So if you break the Vault config while the client is down, the task is going to fail to restore

Oh, good point. But I actually think that, if we don't return an error on missing key here, the task _will_ restore but without any Vault integration 🤔 

I can quickly test this, but I'm now thinking we may need to be able to differentiate between missing and disabled.

Reply:
This took a bit longer than I expected because I started testing several failure modes 😅 

But the results are kind of interesting. As it currently is, the task will indeed fail on restore _but_ that's because of the `vault_hook` failing to get a client here:
https://github.com/hashicorp/nomad/blob/16d71582f68ba5a74504ae6311fe8ac6c8eaea9b/client/allocrunner/taskrunner/vault_hook.go#L166-L169

<img width="836" alt="image" src="https://github.com/hashicorp/nomad/assets/775380/251f36e5-8437-4078-8b10-e77d405ab6a4">

The task failing and being rescheduled in a node with the required Vault cluster makes sense, but if it weren't for the `vault_hook` the task would try to start and just hang for a while trying to get the secret:

<img width="1223" alt="image" src="https://github.com/hashicorp/nomad/assets/775380/eaf72f91-e2e1-454c-8faf-22a0738ce758">

For some reason, after 6m (kind of a random value 🤔), consul-template tries to render the template again, but fails. The interesting bit here is that it seems to be using some default value for the Vault address.

So while technically correct right now, I feel like hooks should work (and fail) independently, regardless of other hooks or the order in which they're called.

All this leads me to think that I _should_ raise an error here, meaning that I would need to access the map of Vault clients directly instead of using `GetVaultConfig()`. Or maybe `GetVaultConfig()` should have an `ok` return as well?

Any second opinions on this?

Reply:
> So while technically correct right now, I feel like hooks should work (and fail) independently, regardless of other hooks or the order in which they're called.

I don't know about that... many or most of the hooks are dependent on ordering (for the prerun method). It's kind of implicit in the design that each sets up the environment needed for those that follow (ex. allocdir before identity, identity before Consul/Vault, Consul/Vault tokens before templates or services, CSI before volume mounts, etc.)

That being said, I wouldn't be opposed to raising an error here. It lets us bail out from a situation we know we'll never properly recover from as early as possible. We can probably just check for a nil config from `GetVaultConfigs` and then return an error like `vault cluster "foo" does not exist or is not enabled` and that'll avoid having to return something more complex than the map we already have.

Reply:
> I don't know about that... many or most of the hooks are dependent on ordering (for the prerun method). It's kind of implicit in the design that each sets up the environment needed for those that follow 

Ah good point, they're called in a specific order on purpose. 

Interestingly, passing the whole Vault config to the template manager (as described in https://github.com/hashicorp/nomad/pull/18790#discussion_r1364436792) helps us solve this problem as well since the `template_hook` can better detect the error, so we get a two-for-one fix 😄